packages/core/src/render3/di_setup.ts

```diff
@@ -74,10 +76,22 @@ function resolveProvider(
   } else {
     const tView = getTView();
     const lView = getLView();
+    const tNode = getCurrentTNode()!;
     let token: any = isTypeProvider(provider) ? provider : resolveForwardRef(provider.provide);
-    let providerFactory: () => any = providerToFactory(provider);
 
-    const tNode = getCurrentTNode()!;
+    let providerFactory: () => any;
+    if (typeof ngDevMode === 'undefined' || ngDevMode) {
+      const nodeInjector =
+          new NodeInjector(tNode as TElementNode | TContainerNode | TElementContainerNode, lView);
+      const prevInjectContext = setDebugInjectContext(
+          {injector: new DebugNodeInjector(nodeInjector), token: token as Type<unknown>});
+      providerFactory = providerToFactory(provider);
+      emitProviderConfiguredEvent(provider);
+      setDebugInjectContext(prevInjectContext);
+    } else {
+      providerFactory = providerToFactory(provider);
+    }
```

Comment:
In cases like this (where the call to instantiate a provider is close to the "wrapping" code to establish debug DI context), it'd be great to extract out all the logic into a separate fn (within `injector_profiler.ts`), so that it's easier to follow and easier to separate what is dev mode and what is not. Here is a quick example (name of the function and arguments - TBD, this is just to illustrate the idea):
```suggestion
    if (typeof ngDevMode === 'undefined' || ngDevMode) {
      // Create a factory in a debug context in dev mode to provide
      // more information to the injector profiler (which works in dev mode only).
      runInDebugInjectContext(tNode, lView, token, provider, () => {
        providerFactory = providerToFactory(provider);
      });
    } else {
      providerFactory = providerToFactory(provider);
    }
```

Reply:
I'd really love to see this pattern in all other cases as well (and there might be separate fns for NodeInjector and Module/EnvironmentInjector), but we can do it as a followup as well.